sdk/storage/azure-storage-blob/azure/storage/blob/_shared/encryption.py

```diff
@@ -115,23 +158,244 @@ def __init__(self, content_encryption_IV, encryption_agent, wrapped_content_key,
         '''
 
-        _validate_not_none('content_encryption_IV', content_encryption_IV)
         _validate_not_none('encryption_agent', encryption_agent)
         _validate_not_none('wrapped_content_key', wrapped_content_key)
 
+        # Validate we have the right matching optional parameter for the specified algorithm
+        if encryption_agent.encryption_algorithm == _EncryptionAlgorithm.AES_CBC_256:
+            _validate_not_none('content_encryption_IV', content_encryption_IV)
+        elif encryption_agent.encryption_algorithm == _EncryptionAlgorithm.AES_GCM_256:
+            _validate_not_none('encrypted_region_info', encrypted_region_info)
+        else:
+            raise ValueError("Invalid encryption algorithm.")
+
         self.content_encryption_IV = content_encryption_IV
+        self.encrypted_region_info = encrypted_region_info
         self.encryption_agent = encryption_agent
         self.wrapped_content_key = wrapped_content_key
         self.key_wrapping_metadata = key_wrapping_metadata
 
 
-def _generate_encryption_data_dict(kek, cek, iv):
+class GCMBlobEncryptionStream:
+    """
+    A stream that performs AES-GCM encryption on the given data as
+    it's streamed. Data is read and encrypted in regions. The stream
+    will use the same encryption key and will generate a guaranteed unique
+    nonce for each encryption region.
+    """
+    def __init__(
+            self,
+            content_encryption_key: bytes,
+            data_stream: BinaryIO,
+            ):
+        """
+        :param content_encryption_key: The encryption key to use.
+        :param data_stream: The data stream to read data from.
+        """
+        self.content_encryption_key = content_encryption_key
+        self.data_stream = data_stream
+
+        self.offset = 0
+        self.current = b''
+        self.nonce_counter = 0
+
+    def read(self, size: int = -1) -> bytes:
+        """
+        Read data from the stream. Specify -1 to read all available data.
+
+        :param size: The amount of data to read. Defaults to -1 for all data.
+        """
+        result = BytesIO()
+        remaining = sys.maxsize if size == -1 else size
+
+        while remaining > 0:
+            # Start by reading from current
+            if len(self.current) > 0:
+                read = min(remaining, len(self.current))
+                result.write(self.current[:read])
+
+                self.current = self.current[read:]
+                self.offset += read
+                remaining -= read
+
+            if remaining > 0:
+                # Read one region of data and encrypt it
+                data = self.data_stream.read(_GCM_REGION_DATA_LENGTH)
+                if len(data) == 0:
+                    # No more data to read
+                    break
+
+                self.current = self._encrypt_region(data)
+
+        return result.getvalue()
+
+    def _encrypt_region(self, data: bytes) -> bytes:
```

Comment:
Possibly missing a `:param bytes data:`?

Reply:
I had left it out since this is a "private" method. Will add it anyway.